### PR TITLE
Follow/Unfollow from profile pages

### LINKED ISSUE #10: Implement Follow & Favorite buttons as views for turbo-frames
Render (Follow Username) and (Favorite Post) buttons in separate views that can be lazily loaded into turbo-frames.

i.e:
_/username/follow_ 
_/article/slug/favorite_

Render views for logged out and logged in states. Logged out states could just show as "# of followers" and "favorite count" instead of buttons.

### PR DESCRIPTION
**Pull request by davish**
Closes #10 and #9.

This PR will allow for follow/unfollow from profile pages using turbo frames. The "level-up" to frames might have been a bit premature, but this will let us also add the follow button easily to article details when that's merged in.

### CODE REVIEW

**Review by C4ptainCrunch — COMMENTED** [review]
Shouldn't the POST redirect to the profile page directly ? This would spare us the need for a view and should work as is because the profile page already contains the turbo-frame. This would also mean that even without turbo the form would still work :)

**Comment by davish** [comment]
> Shouldn't the POST redirect to the profile page directly ? This would spare us the need for a view and should work as is because the profile page already contains the turbo-frame. This would also mean that even without turbo the form would still work :)

I was thinking about this, and for the demo it definitely can work to redirect to the full page rather than the partial. I do think we'll still need a view to handle the form action, since there's already a POST for updating a profile. The one thing I did want to demonstrate was the independent cacheing that DHH et al talk about a lot for HEY. Basically using the lazy-loaded frames as a way to load in user-specific content (in this case, whether the logged-in user has currently followed or not).

I think that at this point, the idioms are still developing around frames. Should the default be to work 100% with no JS and always return full pages? Or can some routes return more bare-bones partials for lazy loading? Personally, I think that our goal shouldn't be to work fully with no JS/no Turbo for every feature, since I feel like that restricts the kind of things you can do.

File: profiles/views.py [hunk 1]
```diff
@@ -36,4 +38,26 @@ def edit(request):
...
+    }
+
+    if request.method == "GET":
+        return render(request, "profile/_follow.html", context)
+    elif request.method == "POST":
+        if is_following:
+            request.user.profile.unfollow(target_user.profile)
+        else:
+            request.user.profile.follow(target_user.profile)
+        return HttpResponseRedirect(reverse('follow_profile', kwargs={"profile": profile}), status=303)
```
Comment (blopker) [inline profiles/views.py]: Nice! I might add another `request.user.is_authenticated` for the POST though:

```python
def follow(request, profile):
    target_user = get_object_or_404(get_user_model(), username=profile)
    if request.user.is_authenticated:
        is_following = request.user.profile.is_following(target_user.profile)
    else:
        is_following = False

    context = {
        "target_user": target_user,
        "is_following": is_following,
    }

    if request.method == "POST" and request.user.is_authenticated:
        if is_following:
            request.user.profile.unfollow(target_user.profile)
        else:
            request.user.profile.follow(target_user.profile)
        return HttpResponseRedirect(reverse('follow_profile', kwargs={"profile": profile}), status=303)
    
    return render(request, "profile/_follow.html", context)
```

Comment (blopker) [reply profiles/views.py]: Do we need the 303 in a frame?

File: realworld/urls.py [hunk 2]
```diff
@@ -26,6 +26,7 @@
     path("editor/<slug:slug>/", articles_views.EditArticle.as_view(), name="article_edit"),
     path("editor/", articles_views.CreateArticle.as_view(), name="article_create"),
     path("@<slug:profile>/", profile_views.view, name="profile_view"),
+    path("@<slug:profile>/follow/", profile_views.follow, name="follow_profile"),
```
Comment (blopker) [inline realworld/urls.py]: Might name it `profile_follow` just to keep the namespacing consistent.

File: templates/profile/_follow.html [hunk 3]
```diff
@@ -0,0 +1,17 @@
+<turbo-frame id="follow-button-{{ target_user.username }}">
+    {% if request.user.is_authenticated %}
+        <form method="POST" action="{% url 'follow_profile' target_user.profile %}">
+            {% csrf_token %}
+            <button class="btn btn-sm btn-outline-secondary action-btn">
```
Comment (blopker) [inline templates/profile/_follow.html:5]: Do you need to change/add/remove a class here based on the follow status?

File: templates/profile/detail.html [hunk 4]
```diff
@@ -7,10 +7,10 @@
...
           <h4>{{ article_user.username }}</h4>
           <p>{{ article_user.profile.bio }}</p>
-          <button class="btn btn-sm btn-outline-secondary action-btn">
-            <i class="ion-plus-round"></i>
-            &nbsp; Follow {{ article_user.username }}
-          </button>
+          <turbo-frame
+                  id="follow-button-{{ article_user.username }}"
+                  src="{% url 'follow_profile' article_user.username %}"
+          />
```
Comment (blopker) [inline templates/profile/detail.html:13]: Would be great to get a standard loading spinner for this site's frames. Maybe in another PR though.

**Comment by blopker** [comment]
RE full page vs partial: in this case the follow button also shows up on other pages as well. If we wanted it as a standalone page I think we'd want a page that only had the follow button on it, which isn't that useful. I think the follow button just doesn't stand on its own so a partial is a good option. If it was an actual form that could be used on its own then the page approach seems worthwhile. Just some thoughts.

File: profiles/views.py (hunk 1 above)
Comment (davish) [reply profiles/views.py]: 303 isn't strictly necessary, we could just render the form again, or just return a default 302. For consistency's sake, though, I kind of like the 303. But it is an extra request compared to just rendering again.

File: templates/profile/_follow.html (hunk 3 above)
Comment (davish) [reply templates/profile/_follow.html:5]: Looking at the [starter kit templates](https://github.com/gothinkster/realworld-starter-kit/blob/master/FRONTEND_INSTRUCTIONS.md), there doesn't seem to be any example the subscribed state for the button. [demo.realworld.io](https://demo.realworld.io) doesn't seem to have any difference either, besides "Follow/Unfollow"

File: templates/profile/detail.html (hunk 4 above)
Comment (davish) [reply templates/profile/detail.html:13]: I'll open an issue for it!

File: templates/profile/_follow.html (hunk 3 above)
Comment (blopker) [reply templates/profile/_follow.html:5]: On https://demo.realworld.io it does toggle the class from `btn-outline-secondary` to `btn-secondary`. Not a big deal though.

**Comment by luanfonceca** [comment]
> RE full page vs partial: in this case the follow button also shows up on other pages as well. If we wanted it as a standalone page I think we'd want a page that only had the follow button on it, which isn't that useful. I think the follow button just doesn't stand on its own so a partial is a good option. If it was an actual form that could be used on its own then the page approach seems worthwhile. Just some thoughts.

Agree, I just did that on #19, for two main reasons I can think of:
1. That way just looked natural to me with the almost-zero knowledge on turbo-frames
2. I was influenced directly by this PR haha

It really looks a lot like a Partial to me. In my case, the Favorite action will redirect the user to the Article page anyway, so it was just natural to make it that way.

If I do a full page reload, I would get ride of the state issue I'm facing right now, here's the scenario:
1. I have two Favorite buttons at the same page
2. We a user favorites a article
3. Only the turbo-frame they clicked gets updated to reflect the new state.

**Comment by blopker** [comment]
So where did we land on this PR? Do all the buttons update when one is pressed? The page reload approach doesn't seem that bad especially if the scroll position is preserved.

**Comment by davish** [comment]
@blopker This PR only has one button per page, so not an issue. I think the above comment was referring to #19. This PR is ready for merge with an approving review!

**Review by blopker — APPROVED** [review]

**Pull request merged**